code/modules/overmap/ftl_shunt/core.dm

```diff
@@ -0,0 +1,551 @@
+/obj/machinery/ftl_shunt/core/proc/start_shunt()
+	var/shunt_distance
+	var/vessel_mass = ftl_computer.linked.get_vessel_mass()
+	var/shunt_turf = locate(shunt_x, shunt_y, GLOB.using_map.overmap_z)
+
+	if(stat & (BROKEN))
+		return FTL_START_FAILURE_BROKEN
+	if(stat & (NOPOWER))
+		return FTL_START_FAILURE_POWER
+
+	if(!length(fuel_ports)) //no fuel ports
+		find_ports()
+		if(!length(fuel_ports))
+			return FTL_START_FAILURE_OTHER
+
+	if(world.time <= cooldown)
+		return FTL_START_FAILURE_COOLDOWN
+
+	shunt_distance = get_dist(get_turf(ftl_computer.linked), shunt_turf)
+	required_fuel_joules = (vessel_mass * JOULES_PER_TON) * shunt_distance
+
+	if(required_fuel_joules > get_fuel(fuel_ports))
+		return FTL_START_FAILURE_FUEL
+
+	charge_interval = get_charge_time()
+	charge_time = world.time + charge_interval
+	//If we've gotten to this point then we're okay to start charging up.
+	charging = TRUE
+	charge_started = world.time
+
+	var/eta = round((get_charge_time() / 600))
+
+	var/announcetxt = replacetext(shunt_start_text, "%%TIME%%", "[eta] minutes.")
+
+	if(use_power != POWER_USE_ACTIVE)
+		update_use_power(POWER_USE_ACTIVE)
+
+	if(sabotaged)
+		for(var/mob/living/carbon/human/H in GLOB.living_mob_list_) //Give engineers a hint that something might be very, very wrong.
+			if(!(H.z in ftl_computer.linked.map_z))
+				continue
+			if(H.skill_check(SKILL_ENGINES, SKILL_EXPERT))
+				to_chat(H, SPAN_DANGER("The deck vibrates with a harmonic that sets your teeth on edge and fills you with dread."))
+
+	ftl_announcement.Announce(announcetxt, "FTL Shunt Management System", new_sound = sound('sound/misc/notice2.ogg'))
+	update_icon()
+	return FTL_START_CONFIRMED
+
+//Cancels the in-progress shunt.
+/obj/machinery/ftl_shunt/core/proc/cancel_shunt(var/silent = FALSE)
+	if(!charging) //Not preparing for a jump.
+		return
+	charging = FALSE
+	charge_started = 0
+	charge_time = null
+	cooldown = null
+	required_fuel_joules = null
+	if(!silent)
+		ftl_announcement.Announce(shunt_cancel_text, "FTL Shunt Management System", new_sound = sound('sound/misc/notice2.ogg'))
+	update_use_power(POWER_USE_IDLE)
+	update_icon()
+	chargepercent = 0
+
+//Starts the shunt, and then hands off to do_shunt to finish it.
+/obj/machinery/ftl_shunt/core/proc/execute_shunt()
+	ftl_announcement.Announce(shunt_spooling_text, "FTL Shunt Management System", new_sound = sound('sound/misc/notice2.ogg'))
+	if(sabotaged)
+		cancel_shunt(TRUE)
+		do_sabotage()
+		return
+
+	update_use_power(POWER_USE_IDLE)
+	var/destination = locate(shunt_x, shunt_y, GLOB.using_map.overmap_z)
+	var/jumpdist = get_dist(get_turf(ftl_computer.linked), destination)
+	var/obj/effect/portal/wormhole/W = new(destination) //Generate a wormhole effect on overmap to give some indication that something is about to happen.
+	QDEL_IN(W, 6 SECONDS)
+	addtimer(CALLBACK(src, .proc/do_shunt, shunt_x, shunt_y, jumpdist, destination), 6 SECONDS)
+	jumping = TRUE
+	update_icon()
+	for(var/mob/living/carbon/M in world)
```

Comment:
You marked afterthought's comment as resolved without changing this or responding.